### ISSUE TITLE
Socket Error with multiple SRV record entries in 2.4.0-2.4.2

### ISSUE THREAD

**Issue by lseelenbinder** [haproxy/haproxy#1270]
## Detailed description of the problem

We use SRV records + `server-template` with `ipv6@` prepended to the SRV record to manage our backends. This worked fine on latest 2.2.x. Upgrading to 2.4.0 (Debian package), we're seeing only the _first_ server come up. The other fail with a socket error. Thus if the SRV record only has a single entry, the error doesn't manifest. If it has multiple, the others are always marked down.

One time the logs included this message (it's not consistent):

`Cannot bind to source address before connect() for backend …. Aborting.`

Relevant config:

```
  server-template server        4 ipv6@<SRV>               crt <crt-path> ca-file <ca-path> verifyhost <verify-host>
```

## Expected behavior

All servers are available.

## Steps to reproduce the behavior

1. Unknown since an upgrade likely invalidated any assumptions about if the config was correct. 😄 

## Do you have any idea what may have caused this?

Upgrading.

## Do you have an idea how to solve the issue?

Downgrading. 😉 

## What is your configuration?

```
(paste your output here)
```

## Output of `haproxy -vv` and `uname -a`

```
HAProxy version 2.4.0-1~bpo10+1 2021/05/21 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.0.html
Running on: Linux 5.11.9-x86_64-linode142 #1 SMP Wed Mar 24 11:28:11 EDT 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.4.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

```
Linux hostname 5.11.9-x86_64 #1 SMP Wed Mar 24 11:28:11 EDT 2021 x86_64 GNU/Linux
```

**Comment by bedis** [comment]
Hi @lseelenbinder , 

can you share the resolvers and relevant part of the backend section?
What type of DNS server are you using?

**Comment by EmericBr** [comment]
I was unable to reproduce using current master and dnsmasq.

**Comment by EmericBr** [comment]
This bug seems not related to srv records butt to bind errors or unreachable network, could you provide additionnal information about your haproxy's and networking configuration
.

**Comment by lseelenbinder** [comment]
Thanks @bedis / @EmericBr for confirming it's not directly related to SRV records. I'll dig deeper—I may have some config cruft left from previous iterations that's causing issues now.

Here's the `resolvers` config:

```
resolvers default
  nameserver g1 [2001:4860:4860::8888]:53
  nameserver g2 [2001:4860:4860::8844]:53
  nameserver cf1 [2606:4700:4700::1111]:53
  nameserver cf2 [2606:4700:4700::1001]:53
  accepted_payload_size 8192
  resolve_retries 4

  hold valid      60s
  hold obsolete   30s
  hold timeout    1s

  timeout resolve 5s
  timeout retry   1s
```

I'm using public DNS servers. The records themselves are resolving (each of the servers created by `server-template` that should have an IP address, has an IP address), but only the first one is actually `UP` and connected (the others that do have IP addresses are failing with socket errors).

Backend is quite simple:

```
backend <backend>
  option httpchk GET /health

  # Allow 10k aggregate connections
  fullconn 10000

  server-template server        4 ipv6@<SRV>               crt <crt-path> ca-file <ca-path> verifyhost <verify-host>
  server-template server-bak        4 ipv6@<SRV-backup>               crt <crt-path> ca-file <ca-path> verifyhost <verify-host>
```

Defaults are also quite simple:

```
defaults
  log  global

  default-server ca-file ca-certificates.crt resolvers default inter 5s fastinter 2s downinter 10s init-addr libc,last,none check ssl check-alpn http/1.1 pool-purge-delay 60s max-reuse 1500 alpn http/1.1
  load-server-state-from-file global

  mode  http
  balance roundrobin
  option http-keep-alive
  option allbackups

  http-reuse always
  retry-on all-retryable-errors
  option redispatch 1
  retries 2

  option httplog
  option dontlognull
  option abortonclose

  # connect [to a backend server]
  timeout connect      12400ms
  # wait for a full request [from a client]
  timeout http-request 9300ms
  # wait for any data from the server (time to first byte/header)
  timeout server          60s
  # wait for the client (keep-alive for h1.1, client for h1.1 & h2)
  timeout http-keep-alive  3m
  timeout client           3m

  errorfile 400 /etc/haproxy/errors/400.http
  errorfile 403 /etc/haproxy/errors/403.http
  errorfile 408 /etc/haproxy/errors/408.http
  errorfile 500 /etc/haproxy/errors/500.http
  errorfile 502 /etc/haproxy/errors/502.http
  errorfile 503 /etc/haproxy/errors/503.http
  errorfile 504 /etc/haproxy/errors/504.http
```

**Comment by capflam** [comment]
@lseelenbinder it is strange. In your config snippet there is no "source" parameter anywhere. But the error message you reported is emitted when HAProxy fails to set the source address for an outgoing connection.  In this case, it may be because you try to set an IPv4 source address for an IPv6 outgoing connection. But 2.2.X and 2.40 work the same way on this point. So it may be unrelated with your issue.

could you provide ""show servers state" command output, for both the 2.2.X and the 2.4.0 ? It may be an issue with the health-checks.

**Comment by capflam** [comment]
@lseelenbinder the `ipv6@` prefix followed by a FQDN is ignored by the resolver. I don't know if it is expected or not and I don't know if it is relevant in your case. But you may try to add `resolve-prefer ipv6` on your server lines.

**Comment by lseelenbinder** [comment]
Thanks for the additional information, @capflam!

> could you provide ""show servers state" command output, for both the 2.2.X and the 2.4.0 ? It may be an issue with the health-checks.

I'll get this to you ASAP, but I'm OOO this week, so it will likely be earliest next week.

> the ipv6@ prefix followed by a FQDN is ignored by the resolver. I don't know if it is expected or not and I don't know if it is relevant in your case. But you may try to add resolve-prefer ipv6 on your server lines.

This isn't expected, no. I'll try that line.

>  In your config snippet there is no "source" parameter anywhere.

I originally had one, but removed it (because it is no longer necessary). The socket error persisted in both cases, but the log did not include the pasted line every time.

> In this case, it may be because you try to set an IPv4 source address for an IPv6 outgoing connection. But 2.2.X and 2.40 work the same way on this point. So it may be unrelated with your issue.

All explicitly set IPs in our setup would be IPv6, and, in this case, there's no way the first server from the SRV results would function if it were using an IPv4 (due to firewalls).

It could very well be a health-check related issue, but the most confusing part about it is that not all servers fail—the first in the set works and the rest fail. That may be a useful thread to investigate?

**Comment by capflam** [comment]
> I'll get this to you ASAP, but I'm OOO this week, so it will likely be earliest next week.

No pb, thanks !

> It could very well be a health-check related issue, but the most confusing part about it is that not all servers fail—the first in the set works and the rest fail. That may be a useful thread to investigate?

For now, I don't know. But the health-checks must not be excluded for our investigation scope.

**Comment by lseelenbinder** [comment]
Was finally able to get that output for you:

```
2.2.14-1

> show servers state
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state
srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord
…snip…
8 backend 1 server1 <ip-prefix>:6a06 2 0 1 1 160664 15 3 4 6 0 0 0 server-1 8443 _srv.record.tld
8 backend 2 server2 <ip-prefix>:6a3b 2 0 1 1 625727 15 3 4 6 0 0 0 server-2 8443 _srv.record.tld
8 backend 3 server3 <ip-prefix>:6a14 2 0 1 1 153497 15 3 4 6 0 0 0 server-3 8443 _srv.record.tld
8 backend 4 server4 :: 0 0 1 1 1279935 8 2 0 6 0 0 0 - 0 _srv.record.tld
8 backend 5 server-backup1 <ip-prefix>:6a44 2 0 1 1 1165088 15 3 4 6 0 0 0 server-bak-1 8443 _srv_backup.record.tld
8 backend 6 server-backup2 <ip-prefix>:6a88 2 0 1 1 1164858 15 3 4 6 0 0 0 server-bak-2 8443 _srv_backup.record.tld
8 backend 7 server-backup3 <ip-prefix>:6a01 2 0 1 1 1121463 15 3 4 6 0 0 0 server-bak-3 8443 _srv_backup.record.tld
8 backend 8 server-backup4 :: 0 0 1 1 1279935 8 2 0 6 0 0 0 - 0 _srv_backup.record.tld


2.4.0

> show servers state
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state
srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
…snip…
8 backend 1 server1 <ip-prefix>:6a06 2 0 1 1 160837 15 3 4 6 0 0 0 server-1 8443 _srv.record.tld 1 0 - - 0
8 backend 2 server2 <ip-prefix>:6a3b 0 0 1 1 17 5 2 0 6 0 0 0 server-2 8443 _srv.record.tld 1 0 - - 0
8 backend 3 server3 <ip-prefix>:6a14 0 0 1 1 17 5 2 0 6 0 0 0 server-3 8443 _srv.record.tld 1 0 - - 0
8 backend 4 server4 :: 0 0 1 1 1280108 8 2 0 6 0 0 0 - 0 _srv.record.tld 1 0 - - 0
8 backend 5 server-backup1 <ip-prefix>:6a44 2 0 1 1 1165261 15 3 4 6 0 0 0 server-bak-1 8443 _srv_backup.record.tld 1

0 backend
8 backend 6 server-backup2 <ip-prefix>:6a88 0 0 1 1 14 5 2 0 6 0 0 0 server-bak-2 8443 _srv_backup.record.tld 1 0 - -

0
8 backend 7 server-backup3 <ip-prefix>:6a01 0 0 1 1 14 5 2 0 6 0 0 0 server-bak-3 8443 _srv_backup.record.tld 1 0 - -

0
8 backend 8 server-backup4 :: 0 0 1 1 1280108 8 2 0 6 0 0 0 - 0 _srv_backup.record.tld 1 0 - - 0
```

I fiddled around with health check settings, adjusted the `ipv6@` prefix, and nothing made any difference. One thing I observed is the socket error failures have a check duration that is as long as a normal health check, so it's not *just* trying to connect and failing (it's a few multiples of the ping latency, so it seems to be at least trying to do a SSL handshake, too).

**Comment by capflam** [comment]
Thanks @lseelenbinder, I'll take a look.

**Comment by capflam** [comment]
The 2.4.0 output is a bit mangled but there is nothing strange. And indeed, only the first server is UP. Others are DOWN. Same for backup servers. I'm pretty sure the socket error is reported by the health-checks. Could you enable the check traces on the 2.4.0 ? Because it is pretty verbose , you can start by enabling the errors only with the following command on the CLI:
```
(echo "trace check sink buf0; trace check level error; trace check verbosity complete; trace check start now;  show events buf0 -w" ; cat) | socat {HAPROXY-SOCKET} -
```

**Comment by lseelenbinder** [comment]
Will get that for you ASAP @capflam.

**Comment by lseelenbinder** [comment]
@capflam finally had the chance to do this:

```
$ (echo "trace check sink buf0; trace check level error; trace check verbosity complete; trace check start now;  show events buf0 -w" ; cat) | sudo socat /run/haproxy/admin.sock -




<0>2021-06-25T06:07:12.509884+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server-backup4 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b99ec8(0x00000007) - conn=0x560d61676e70(0x10102300) cs=0x560d6145eca0(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:12.509905+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server-backup4 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b99ec8(0x00000007) - conn=0x560d61676e70(0x10102300) cs=0x560d6145eca0(0x00000100)
<0>2021-06-25T06:07:14.110969+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server-backup3 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b98d08(0x00000007) - conn=0x560d61676e70(0x10102300) cs=0x560d6145eca0(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:14.111029+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server-backup3 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b98d08(0x00000007) - conn=0x560d61676e70(0x10102300) cs=0x560d6145eca0(0x00000100)
<0>2021-06-25T06:07:14.526985+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server-backup2 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b97488(0x00000007) - conn=0x560d613a0610(0x101c2300) cs=0x560d614d4000(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:14.527006+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server-backup2 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b97488(0x00000007) - conn=0x560d613a0610(0x101c2300) cs=0x560d614d4000(0x00000100)
<0>2021-06-25T06:07:15.014891+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server4 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b94478(0x00000007) - conn=0x560d61617b00(0x10102300) cs=0x560d6144a7e0(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:15.014915+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server4 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b94478(0x00000007) - conn=0x560d61617b00(0x10102300) cs=0x560d6144a7e0(0x00000100)
<0>2021-06-25T06:07:18.489033+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server3 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b93268(0x00000007) - conn=0x560d615c6a90(0x10102300) cs=0x560d614d44f0(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:18.489054+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server3 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b93268(0x00000007) - conn=0x560d615c6a90(0x10102300) cs=0x560d614d44f0(0x00000100)
<0>2021-06-25T06:07:18.587282+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server2 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b91a38(0x00000007) - conn=0x560d615c6a90(0x10102300) cs=0x560d614d44f0(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:18.587293+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server2 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b91a38(0x00000007) - conn=0x560d615c6a90(0x10102300) cs=0x560d614d44f0(0x00000100)
<0>2021-06-25T06:07:23.017289+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server-backup4 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b99ec8(0x00000007) - conn=0x560d61603e10(0x10102300) cs=0x560d613a1b80(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:23.017311+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server-backup4 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b99ec8(0x00000007) - conn=0x560d61603e10(0x10102300) cs=0x560d613a1b80(0x00000100)
<0>2021-06-25T06:07:24.381848+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server-backup3 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b98d08(0x00000007) - conn=0x560d61462360(0x10102300) cs=0x560d61615960(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:24.381876+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server-backup3 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b98d08(0x00000007) - conn=0x560d61462360(0x10102300) cs=0x560d61615960(0x00000100)
<0>2021-06-25T06:07:24.593597+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server-backup2 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b97488(0x00000007) - conn=0x560d61615ad0(0x10102300) cs=0x560d61416290(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:24.593607+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server-backup2 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b97488(0x00000007) - conn=0x560d61615ad0(0x10102300) cs=0x560d61416290(0x00000100)
<0>2021-06-25T06:07:25.432720+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server4 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b94478(0x00000007) - conn=0x560d61615ad0(0x10102300) cs=0x560d6148e680(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:25.432736+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server4 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b94478(0x00000007) - conn=0x560d61615ad0(0x10102300) cs=0x560d6148e680(0x00000100)
<0>2021-06-25T06:07:28.074533+00:00 [00|check|0|pcheck.c:2260] report connection error : [H] SRV=server3 status=0/2 DOWN - last=SOCKERR(5)/UNK(0) check=0x560d60b93268(0x00000007) - conn=0x560d61615ad0(0x10102300) cs=0x560d614d4050(0x00000100) - tcp-check=(HTTP,1)
<0>2021-06-25T06:07:28.074553+00:00 [00|check|0|/check.c:1043] report connection error : [H] SRV=server3 status=0/2 DOWN - last=SOCKERR(5)/FAIL(2) check=0x560d60b93268(0x00000007) - conn=0x560d61615ad0(0x10102300) cs=0x560d614d4050(0x00000100)
```

Hope it helps! (By the way, this is on 2.4.1 now, so no fixes in that release addressed this.)

**Comment by capflam** [comment]
Thanks. So the socket error is indeed reported by the health check, when data are sent or received, after the connection establishment. The error was detected at the mux level (error reported on the connection and on the conn-stream). It is a bit hard to know exactly what happens for now. It may be a shutdown detected too early.

Sorry but I will ask you other traces. It is boring but I have no other way to spot the bug. You may try to get errors from the H1 multiplexer, because the error is triggered at this level:

```
(echo "trace h1 sink buf0; trace h1 level error; trace h1 verbosity complete; trace h1 start now;  show events buf0 -w -n" ; cat) | socat {HAPROXY-SOCKET} -
```

And eventually, it could be helpful to have a full traces of a failing health-check. But this will be verbose. If you prefer, you may wait my analysis of the H1 mux traces.

```
(echo "trace check sink buf0; trace check level developer; trace check verbosity complete; trace check start now;  show events buf0 -w -n" ; cat) | socat {HAPROXY-SOCKET} -
```

Note that you can stop all traces with `echo "trace 0" | socat {HAPROXY-SOCKET} -`

**Comment by lseelenbinder** [comment]
Hi @capflam, took a bit, but I was able to capture these traces.

The first trace didn't result in anything (I was testing 2.4.2, so perhaps a bug has been fixed since then); but the servers definitely weren't staying up, as demonstrated by the second trace.

I emailed it directly, since it's quite verbose and has some potentially sensitive information included.

**Comment by capflam** [comment]
Thanks a lot, I'll take a look.

**Comment by capflam** [comment]
If no error are reported in the H1 trace, it means the error is not emitted by the mux itself. For instance, it is not an HTTP parsing error. However, a connection error is reported. CO_FL_ERROR and CS_FL_ERROR flags are set on the connection and the conn-stream, respectively, when the error is detected by the health-check task. On the connection, no shutdown is reported. CO_FL_SOCK_RD_SH and CO_FL_SOCK_WR_SH are not set. So it is not an error at the connection level too. It is most probably a SSL error. The handshake is finished and the connection is fully established when it happens, thus, if I'm right, the error is detected during a SSL_read or a SSL_write.

For now, I have no idea why this happens and why not on the first server but all the others. I will investigate in that direction and let you know if I found something.

**Comment by capflam** [comment]
@lseelenbinder, I'm unable to reproduce the bug. However, there was a recent bug report about SSL that may explain your issue. A fix was pushed in upstream (0498fa4). It may be applied on top of the 2.4.2. Could you try it ? If it is not possible, the fix will be backported very soon.

**Comment by lseelenbinder** [comment]
Thanks, @capflam! That actually makes a lot of sense. We do use `default-server` with a client cert. I'll try again when a new release is cut, and let you know how it goes.

**Comment by capflam** [comment]
FYI, the fix was backported to 2.4.

**Comment by lseelenbinder** [comment]
Hi @capflam, looks like whatever the issue _was_, it's been resolved in 2.4.3! Thanks for your help on this. 👍
